File: services/api/src/aec_api/work_queue.py
"""R26-WORK-QUEUE — what is in your court right now, in the order it needs you, actionable in place.

The **Work** room's whole job. The platform already knew both halves — `my_work` finds the records
assigned to you or waiting on your party, and `due_feed` knows what is late — but they were separate
lists on separate screens, and neither carried the one thing that turns a list into a queue: *what
you can do about each item without leaving it*.

Two decisions shape this.

**An item with no due date is not "not urgent".** It is *undated*, which is a different fact and often
a data gap somebody should close. Folding it into "later" would sort a genuinely urgent, dateless RFI
below a routine item due next month, and would hide the gap that caused it. `undated` is its own
bucket, named, sitting above `later` — because an item nobody dated still needs a decision, and the
first decision is usually "when is this due".

**Every item carries the actions the ENGINE will honour.** Not "the actions this module defines" —
the ones this caller's party can actually perform from this state. A queue that offers a button the
API then rejects with a 409 is worse than one that offers nothing, because it spends the user's trust
before it spends their time.
"""
from __future__ import annotations

from datetime import date, timedelta
from typing import Any

from sqlalchemy.orm import Session

#: Buckets, in the order they should be worked. `undated` sits ABOVE `later` on purpose.
BUCKETS = ["overdue", "today", "soon", "undated", "later"]

#: Days ahead that still counts as "soon".
SOON_DAYS = 7

BUCKET_LABEL = {
    "overdue": "Overdue",
    "today": "Due today",
    "soon": f"Next {SOON_DAYS} days",
    "undated": "No due date",
    "later": "Later",
}

#: Why each bucket exists, so a UI does not have to invent the explanation.
BUCKET_MEANS = {
    "overdue": "Past its due date",
    "today": "Due before the end of today",
    "soon": f"Due within {SOON_DAYS} days",
    "undated": "Nobody has set a due date — that is a gap, not a low priority",
    "later": "Dated, and beyond the horizon",
}


def bucket(due: str | None, today: date | None = None, soon_days: int = SOON_DAYS) -> str:
    """Which bucket a due date falls in. Absent or unparseable → `undated`, never `later`.

    An unparseable date is the same class of thing as a missing one: the platform does not know when
    this is due. Silently treating it as far-future would bury it.
    """
    if not str(due or "").strip():
        return "undated"
    try:
        d = date.fromisoformat(str(due)[:10])
    except ValueError:
        return "undated"
    t = today or date.today()
    if d < t:
        return "overdue"
    if d == t:
        return "today"
    return "soon" if d <= t + timedelta(days=max(0, soon_days)) else "later"


def _sort_key(item: dict[str, Any]) -> tuple:
    """Within a bucket: dated before undated, earliest first, assigned-to-me ahead of ball-in-court."""
    return (0 if item.get("due") else 1,
            str(item.get("due") or "9999-12-31"),
            0 if item.get("reason") == "assigned" else 1,
            str(item.get("ref") or ""))


def queue(db: Session, project_id: str, user: str, party: str | None,
          today: date | None = None, soon_days: int = SOON_DAYS) -> dict[str, Any]:
    """The caller's work queue: their `my_work` feed, dated, bucketed, and with live actions.

    Built on `my_work` rather than beside it — a second definition of "in my court" is exactly how two
    screens come to disagree, which is the defect `test_consistency` exists to prevent.
    """
    from . import modules as me
    from .modules_query import available_actions

    items = me.my_work(db, project_id, user, party)
    out: list[dict[str, Any]] = []
    for it in items:
        mod = me.get_module(it["module"])
        due = None
        df = me._due_field_name(mod)
        if df:
            try:
                rec = me.get_record(db, it["module"], project_id, it["id"])
                due = (rec.get("data") or {}).get(df)
            except Exception:                 # noqa: BLE001 — a record that vanished is not a crash
                due = None
        acts = available_actions(mod, it["state"], party)
        out.append({**it, "due": due, "bucket": bucket(due, today, soon_days),
                    "actions": [a["action"] for a in acts],
                    # a transition the module gates behind required fields cannot be a one-click
                    # button, and saying so beats offering one that 409s
                    "blocked_actions": [a["action"] for a in acts if a.get("requires")]})

    grouped = {b: sorted([x for x in out if x["bucket"] == b], key=_sort_key) for b in BUCKETS}
    actionable = sum(1 for x in out if x["actions"])
    return {
        "buckets": [{"key": b, "label": BUCKET_LABEL[b], "means": BUCKET_MEANS[b],
                     "count": len(grouped[b]), "items": grouped[b]} for b in BUCKETS],
        "total": len(out),
        "actionable": actionable,
        # An item in your court that you cannot act on is a queue entry with no exit. Counting them
        # separately is the difference between "you have 14 things" and "you have 14 things and 3 of
        # them are waiting on somebody else".
        "no_action": len(out) - actionable,
        "note": ("`undated` is its own bucket, not the bottom of `later` — an item nobody dated needs "
                 "a decision, and the first decision is usually when it is due. Actions are the ones "
                 "this party can perform from this state, so a button offered here will not 409."),
    }

File: apps/web/src/portal/panels/workQueue.ts
import { escapeHtml as esc, toast } from "../../ui/feedback";
import { markPrimary } from "../../ui/a11yJourney";
import type { PanelContext } from "../panelContext";
import type { QueueItem } from "../../api/types";

/**
 * R26-WORK-QUEUE — the Work room's home: what is in your court, in the order it needs you, acted on
 * without leaving the queue.
 *
 * The design decisions that matter are in the engine (`work_queue.py`); the two this file has to
 * honour are:
 *
 * * **`undated` renders as its own group, above `later`, with its reason on screen.** An item nobody
 *   dated is a gap somebody should close, and the fastest way to close it is to be told it exists.
 * * **A button here runs a transition the server has already agreed to.** The engine returns only the
 *   actions this party can perform from this state, so the queue never offers a click that 409s. An
 *   action the module gates behind required fields is shown as a *link into the record* instead —
 *   it needs a form, not a click, and pretending otherwise is how a one-click button becomes a dead
 *   end.
 */
export async function renderWorkQueue(ctx: PanelContext) {
  const pid = ctx.host.projectId()!;
  ctx.root.innerHTML = "";
  ctx.root.appendChild(ctx.bar("✅ My work", () => { ctx.activeKey = null; void ctx.renderHome(); ctx.buildNav(); }));

  const body = document.createElement("div");
  body.innerHTML = `<div class="meta">Gathering what is in your court…</div>`;
  ctx.root.appendChild(body);

  // Open the record's own module, filtered to it — the queue hands off rather than re-implementing
  // a record view, so there is one place a record is edited.
  const open = (it: QueueItem) => {
    const m = ctx.mods.find((x) => x.key === it.module);
    if (m) { ctx.activeKey = it.module; void ctx.openModule(m, { q: it.ref }); ctx.buildNav(); }
  };

  async function load() {
    let q;
    try { q = await ctx.host.api.workQueue(pid); }
    catch (e) { body.innerHTML = `<div class="meta">Couldn't load the queue: ${esc((e as Error).message)}</div>`; return; }
    body.replaceChildren();

    const restoreFocus = () => {
      const p = body.querySelector<HTMLElement>("[data-room-primary]");
      if (p && !body.contains(document.activeElement)) p.focus();
    };

    const head = document.createElement("div");
    head.className = "dash-card"; head.style.marginBottom = "10px";
    // "You have 14 things" and "14 things, 3 of them waiting on somebody else" are different
    // statements, and only the second lets someone plan a day.
    head.innerHTML = `<div class="section-title" style="margin:0 0 4px">In your court</div>`
      + `<div class="meta">${q.total} item${q.total === 1 ? "" : "s"} · `
      + `<b>${q.actionable}</b> you can move now`
      + (q.no_action ? ` · ${q.no_action} waiting on someone else` : "") + `</div>`;
    body.appendChild(head);

    if (!q.total) {
      const none = document.createElement("div");
      none.className = "dash-card";
      none.innerHTML = `<div class="meta">Nothing is in your court right now.</div>`;
      const refresh = document.createElement("button");
      refresh.type = "button";
      refresh.className = "tool-btn";
      refresh.style.marginTop = "8px";
      refresh.textContent = "Refresh queue";
      refresh.onclick = () => void load();
      markPrimary(refresh, "work");
      none.appendChild(refresh);
      body.appendChild(none);
      restoreFocus();
      return;
    }

    let marked = false;
    for (const b of q.buckets) {
      // An empty bucket is skipped EXCEPT overdue: rendering "Overdue 0" is worth the line, because
      // a missing heading makes the reader check whether it was hidden or genuinely clear.
      if (!b.count && b.key !== "overdue") continue;
      const card = document.createElement("div");
      card.className = "dash-card"; card.style.marginBottom = "10px";
      card.dataset.bucket = b.key;
      const h = document.createElement("div");
      h.className = "section-title"; h.style.margin = "0 0 2px";
      h.textContent = `${b.label} · ${b.count}`;
      const why = document.createElement("div");
      why.className = "meta"; why.style.marginBottom = "6px";
      why.textContent = b.means;
      card.append(h, why);

      for (const it of b.items) {
        const row = document.createElement("div");
        row.className = "wq-row";
        row.dataset.module = it.module;

        const main = document.createElement("button");
        main.className = "wq-open";
        main.type = "button";
        main.title = `Open ${it.ref}`;
        if (!marked) { markPrimary(main, "work"); marked = true; }
        main.innerHTML = `<span class="ic">${esc(it.icon || "•")}</span> `
          + `<b>${esc(it.ref)}</b> ${esc(it.title || "(untitled)")}`
          + `<span class="meta"> — ${esc(it.module_name)} · ${esc(it.state)}`
          + (it.due ? ` · due ${esc(it.due)}` : "")
          + (it.reason === "assigned" ? " · assigned to you" : "") + `</span>`;
        main.onclick = () => open(it);
        row.appendChild(main);

        const acts = document.createElement("span");
        acts.className = "wq-acts";
        for (const a of it.actions) {
          if (it.blocked_actions.includes(a)) {
            // Needs a form, not a click. Say which, and go there.
            const link = document.createElement("button");
            link.className = "wq-act wq-act-form";
            link.textContent = `${a}…`;
            link.title = `${a} needs fields filled in — opens the record`;
            link.onclick = () => open(it);
            acts.appendChild(link);
            continue;
          }
          const btn = document.createElement("button");
          btn.className = "wq-act";
          btn.textContent = a;
          btn.title = `${a} — ${it.ref}`;
          btn.onclick = async () => {
            btn.disabled = true;
            try {
              await ctx.host.api.transitionRecord(pid, it.module, it.id, a);
              toast(`${it.ref}: ${a}`, "success");
              await load();                     // the queue is the point — stay in it
            } catch (e) {
              btn.disabled = false;
              toast(`Couldn't ${a} ${it.ref}: ${(e as Error).message}`, "error");
            }
          };
          acts.appendChild(btn);
        }
        if (!it.actions.length) {
          const w = document.createElement("span");
          w.className = "meta"; w.textContent = "waiting on someone else";
          acts.appendChild(w);
        }
        row.appendChild(acts);
        card.appendChild(row);
      }
      body.appendChild(card);
    }
    restoreFocus();
  }

  await load();
}
